### PR TITLE
Support Table and Index partitioning in Babelfish

### PR DESCRIPTION
**Pull request by sumitj824**
### Description
To support partitioned tables and indexes in Babelfish, this commit introduces the following changes:

1. Babelfish stores the creation date and original name of the table in the reloption, whereas PostgreSQL does not allow the usage of reloption for partitioned tables. This commit enables the usage of reloption for Babelfish partitioned tables to store the creation date and original name. This information will be utilized by various system views, such as information_schema_tsql.tables, sys.tables, etc., to display the table creation date and original name.

3. Babelfish supports column ordering and NULLs ordering in table constraints while PostgreSQL does not and to handle dump/restore of this we dump the underlying index of a UNIQUE/PRIMARY KEY constraint and make use of ALTER TABLE ADD CONSTRAINT USING syntax to support column ordering and NULLs ordering in table constraints. PostgreSQL disallows creation of an constraint using index like this for partitioned table so we will need to bypass that check during restore of Babelfish databases.

Signed-off-by: Sumit Jaiswal [sumiji@amazon.com](mailto:sumiji@amazon.com)

Extension PR: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2702
 
### Issues Resolved

Task: BABEL-4968, BABEL-4969
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/access/common/reloptions.c [hunk 1]
```diff
@@ -2001,10 +2003,16 @@ bytea *
 partitioned_table_reloptions(Datum reloptions, bool validate)
 {
 	if (validate && reloptions)
+	{
+		/* For babelfish partitioned table, allow usage of reloptions in TSQL dialect. */
+		if (pltsql_partitioned_table_reloptions_hook && pltsql_partitioned_table_reloptions_hook(reloptions))
```
Comment (Deepesh125) [inline src/backend/access/common/reloptions.c]: check dialect also

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: src/backend/access/common/reloptions.c (hunk 1 above)
Comment (sumitj824) [reply src/backend/access/common/reloptions.c]: Dialect check and dump/restore check are already there within the underlined extension function for the this hook.

**Review by rishabhtanwar29 — APPROVED** [review]

File: src/backend/parser/parse_utilcmd.c [hunk 2]
```diff
@@ -2291,8 +2291,12 @@ transformIndexConstraint(Constraint *constraint, CreateStmtContext *cxt)
...
 							index_name, RelationGetRelationName(heap_rel)),
 					 parser_errposition(cxt->pstate, constraint->location)));
-
-		if (!index_form->indisvalid)
+		/*
+		 * For Babelfish databases, allow the restore of index which is
+		 * marked as not used by queries for partitioned table.
+		 */
+		if (!index_form->indisvalid &&
+			(!bbf_dump_restore || strcmp(bbf_dump_restore, "on") != 0))
```
Comment (Deepesh125) [inline src/backend/parser/parse_utilcmd.c]: We should check if parent rel is partitioned table

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: src/backend/access/common/reloptions.c [hunk 3]
```diff
@@ -2001,10 +2003,16 @@ bytea *
 partitioned_table_reloptions(Datum reloptions, bool validate)
 {
 	if (validate && reloptions)
+	{
+		/* For babelfish partitioned table, allow usage of reloptions in TSQL dialect. */
```
Comment (forestkeeper) [inline src/backend/access/common/reloptions.c]: The comment style should be : 

```
/* 
 * For babelfish partitioned table, allow usage of reloptions in TSQL dialect. 
 */
```

File: src/backend/access/common/reloptions.c (hunk 1 above)
Comment (forestkeeper) [inline src/backend/access/common/reloptions.c]: Should give a more reasonable name for the hook type like : 
```
pltsql_is_partitioned_table_reloptions_allowed_hook
```
the current name looks confusing.

**Review by Deepesh125 — APPROVED** [review]

**Review by KushaalShroff — APPROVED** [review]

**Pull request merged**